toonz/sources/toonzqt/schematicviewer.cpp

```diff
@@ -432,6 +544,198 @@ void SchematicSceneViewer::showEvent(QShowEvent *se) {
   }
 }
 
+//------------------------------------------------------------------
+
+void SchematicSceneViewer::enterEvent(QEvent *e) {
+  switch (m_cursorMode) {
+  case CursorMode::Hand:
+    setToolCursor(this, ToolCursor::PanCursor);
+    break;
+  case CursorMode::Zoom:
+    setToolCursor(this, ToolCursor::ZoomCursor);
+    break;
+  default:
+    setToolCursor(this, ToolCursor::StrokeSelectCursor);
+    break;
+  }
+}
+
+//------------------------------------------------------------------
+
+void SchematicSceneViewer::leaveEvent(QEvent *e) { setCursor(Qt::ArrowCursor); }
+
+//------------------------------------------------------------------
+
+void SchematicSceneViewer::tabletEvent(QTabletEvent *e) {
+  // qDebug() << "[tabletEvent]";
+  if (e->type() == QTabletEvent::TabletPress) {
+    m_stylusUsed = e->pointerType() ? true : false;
+  } else if (e->type() == QTabletEvent::TabletRelease) {
+    m_stylusUsed = false;
+  }
+
+  e->accept();
+}
+
+//------------------------------------------------------------------
+
+void SchematicSceneViewer::gestureEvent(QGestureEvent *e) {
+  // qDebug() << "[gestureEvent]";
+  m_gestureActive = false;
+  if (QGesture *swipe = e->gesture(Qt::SwipeGesture)) {
+    m_gestureActive = true;
+  } else if (QGesture *pan = e->gesture(Qt::PanGesture)) {
+    m_gestureActive = true;
+  }
+  if (QGesture *pinch = e->gesture(Qt::PinchGesture)) {
+    QPinchGesture *gesture = static_cast<QPinchGesture *>(pinch);
+    QPinchGesture::ChangeFlags changeFlags = gesture->changeFlags();
+    QPoint firstCenter = mapFromGlobal(gesture->centerPoint().toPoint());
```

Comment:
I think this is a probable cause of the offset problem.
You are obtaining `firstCenter` by converting `gesture->centerPoint()` with `mapFromGlobal()` , but `gesture->centerPoint()` seems to be already in widget coordinate, not in the global screen coordinate.

Reply:
@shun-iwasawa Something I just noticed on my Windows machine in regards to pinch zooming on my **touchscreen** in both the Scene and Schematic viewers:

When **not** using mapFromGlobal() while getting the firstCenter, if I pinch zoom directly in the center of the either Viewer, it always causes the zoom to shift to/from the left.  I would have to be pinching left of center to zoom without causing a shift left or right.  However if I change the logic to use mapFromGlobal(), pinch zooming at the center does not cause any shifting.

This seems like a minor existing gesture bug to me.  I'd be curious to know if any others with a touchscreen see this pinching behavior also.

With or without mapFromGlobal(), **touchpad** pinching on my Windows machine works exactly the same and correctly, zooming relative to where the pointer is.

Would removing mapFromGlobal() fix the offset on the Mac?  If so, I can put a compile macro around that statement to use mapFromGlobal() for Windows compiles only.